File: internal/protocol/turn_intent.go
package protocol

import (
	"encoding/json"
	"strings"
)

const (
	TurnMetaComposerMode      = "composer_mode"
	TurnMetaContextTier       = "context_tier"
	TurnMetaCanProposeFiles   = "can_propose_files"
	TurnMetaCanRunImplSession = "can_run_impl_session"
	TurnMetaRequiresWorkspace = "requires_workspace"
	TurnMetaGovernance        = "turn_governance"
)

const TurnGovernanceVersion = 1

type TurnGovernance struct {
	Version           int    `json:"version"`
	ComposerMode      string `json:"composer_mode"`
	ContextTier       string `json:"context_tier"`
	TrustPreference   string `json:"trust_preference,omitempty"`
	CanProposeFiles   bool   `json:"can_propose_files"`
	CanRunImplSession bool   `json:"can_run_impl_session"`
	RequiresWorkspace bool   `json:"requires_workspace"`
	Provenance        string `json:"provenance"`
}

func StampTurnGovernance(m *Message, governance TurnGovernance) {
	if m == nil {
		return
	}
	governance.Version = TurnGovernanceVersion
	if m.Metadata == nil {
		m.Metadata = make(map[string]interface{})
	}
	encoded, _ := json.Marshal(governance)
	var value map[string]interface{}
	_ = json.Unmarshal(encoded, &value)
	m.Metadata[TurnMetaGovernance] = value
}

func ExtractTurnGovernance(m *Message) (TurnGovernance, bool) {
	if m == nil || m.Metadata == nil {
		return TurnGovernance{}, false
	}
	raw, ok := m.Metadata[TurnMetaGovernance]
	if !ok {
		return TurnGovernance{}, false
	}
	encoded, err := json.Marshal(raw)
	if err != nil {
		return TurnGovernance{}, false
	}
	var governance TurnGovernance
	if json.Unmarshal(encoded, &governance) != nil || governance.Version != TurnGovernanceVersion {
		return TurnGovernance{}, false
	}
	switch governance.ComposerMode {
	case "ask", "agent", "plan", "export":
	default:
		return TurnGovernance{}, false
	}
	return governance, true
}

// ComposerModeFromMessage returns ask, agent, plan, or export from explicit metadata.
func ComposerModeFromMessage(m *Message) string {
	if governance, ok := ExtractTurnGovernance(m); ok {
		return governance.ComposerMode
	}
	if m == nil || m.Metadata == nil {
		return ""
	}
	if v, _ := m.Metadata[TurnMetaComposerMode].(string); strings.TrimSpace(v) != "" {
		return strings.TrimSpace(v)
	}
	return strings.TrimSpace(m.IdeEditorMode())
}

func (m *Message) IdeEditorModeIsAsk() bool {
	return ComposerModeFromMessage(m) == "ask"
}

func (m *Message) IdeEditorModeIsAgent() bool {
	mode := ComposerModeFromMessage(m)
	return mode == "agent" || mode == ""
}

func (m *Message) IdeEditorModeIsPlan() bool {
	return ComposerModeFromMessage(m) == "plan"
}

func (m *Message) IdeEditorModeIsExport() bool {
	return ComposerModeFromMessage(m) == "export"
}

// TurnCapabilities describes what an agent may do on this message turn.
type TurnCapabilities struct {
	ComposerMode      string
	ContextTier       string
	CanProposeFiles   bool
	CanRunImplSession bool
	RequiresWorkspace bool
}

// ResolveTurnCapabilities derives capabilities from message metadata (UI-first).
func ResolveTurnCapabilities(m *Message) TurnCapabilities {
	if governance, ok := ExtractTurnGovernance(m); ok {
		cap := TurnCapabilities{
			ComposerMode: governance.ComposerMode, ContextTier: governance.ContextTier,
			CanProposeFiles:   governance.CanProposeFiles,
			CanRunImplSession: governance.CanRunImplSession,
			RequiresWorkspace: governance.RequiresWorkspace,
		}
		if cap.ComposerMode == "ask" || cap.ComposerMode == "plan" {
			cap.CanProposeFiles = false
			cap.CanRunImplSession = false
			cap.RequiresWorkspace = false
		}
		if cap.ContextTier == "" {
			cap.ContextTier = "none"
		}
		return cap
	}
	cap := TurnCapabilities{
		ComposerMode: ComposerModeFromMessage(m),
		ContextTier:  ContextScopeFromMessage(m),
	}
	if cap.ComposerMode == "" {
		cap.ComposerMode = "agent"
	}
	switch cap.ComposerMode {
	case "ask", "plan":
		cap.CanProposeFiles = false
		cap.CanRunImplSession = false
	case "export":
		cap.CanProposeFiles = true
		cap.CanRunImplSession = true
		cap.RequiresWorkspace = true
	default: // agent
		cap.CanProposeFiles = true
		cap.CanRunImplSession = m != nil && m.ImplementationSession()
		cap.RequiresWorkspace = cap.CanRunImplSession
	}
	if cap.ContextTier == "" {
		cap.ContextTier = "none"
	}
	return cap
}

// ContextScopeFromMessage reads context_scope metadata.
func ContextScopeFromMessage(m *Message) string {
	if governance, ok := ExtractTurnGovernance(m); ok {
		return governance.ContextTier
	}
	if m == nil || m.Metadata == nil {
		return ""
	}
	s, _ := m.Metadata["context_scope"].(string)
	return strings.TrimSpace(s)
}

File: desktop/src/constants/turnIntent.ts
import type { EffectiveComposerMode } from './composerMode';
import type { ContextScope } from './promptMetadata';

/** Mirrors internal/protocol/turn_intent.go — UI sets metadata; agent resolves capabilities. */
export type TurnCapabilities = {
  composerMode: EffectiveComposerMode;
  contextTier: ContextScope;
  canProposeFiles: boolean;
  canRunImplSession: boolean;
  requiresWorkspace: boolean;
};

export const TURN_META_COMPOSER_MODE = 'composer_mode';
export const TURN_META_CONTEXT_TIER = 'context_tier';
export const TURN_META_CAN_PROPOSE_FILES = 'can_propose_files';
export const TURN_META_CAN_RUN_IMPL_SESSION = 'can_run_impl_session';
export const TURN_META_REQUIRES_WORKSPACE = 'requires_workspace';
export const TURN_META_GOVERNANCE = 'turn_governance';
export const TURN_GOVERNANCE_VERSION = 1;

export function resolveTurnCapabilities(options: {
  composerMode: EffectiveComposerMode;
  contextScope: ContextScope;
  implementationSession?: boolean;
}): TurnCapabilities {
  const { composerMode, contextScope, implementationSession } = options;
  switch (composerMode) {
    case 'ask':
      return {
        composerMode,
        contextTier: contextScope,
        canProposeFiles: false,
        canRunImplSession: false,
        requiresWorkspace: false,
      };
    case 'plan':
      return {
        composerMode,
        contextTier: contextScope,
        canProposeFiles: false,
        canRunImplSession: false,
        requiresWorkspace: false,
      };
    case 'export':
      return {
        composerMode,
        contextTier: contextScope,
        canProposeFiles: true,
        canRunImplSession: true,
        requiresWorkspace: true,
      };
    default:
      return {
        composerMode: 'agent',
        contextTier: contextScope,
        canProposeFiles: true,
        canRunImplSession: Boolean(implementationSession),
        requiresWorkspace: Boolean(implementationSession),
      };
  }
}

/**
 * Attach a versioned turn_governance envelope. Flat can_* keys are omitted —
 * the server stamps canonical governance and capabilities at ingress.
 * composer_mode remains as a compatibility reader for older ingress paths.
 */
export function attachTurnCapabilitiesMetadata(
  metadata: Record<string, unknown>,
  caps: TurnCapabilities,
  options?: { trustPreference?: string }
): Record<string, unknown> {
  const next: Record<string, unknown> = {
    ...metadata,
    [TURN_META_COMPOSER_MODE]: caps.composerMode,
    [TURN_META_GOVERNANCE]: {
      version: TURN_GOVERNANCE_VERSION,
      composer_mode: caps.composerMode,
      context_tier: caps.contextTier,
      trust_preference: options?.trustPreference ?? '',
      can_propose_files: caps.canProposeFiles,
      can_run_impl_session: caps.canRunImplSession,
      requires_workspace: caps.requiresWorkspace,
      provenance: 'desktop_explicit',
    },
  };
  delete next[TURN_META_CONTEXT_TIER];
  delete next[TURN_META_CAN_PROPOSE_FILES];
  delete next[TURN_META_CAN_RUN_IMPL_SESSION];
  delete next[TURN_META_REQUIRES_WORKSPACE];
  return next;
}
